File: runtimes/go/shm_transport.go
package runtime

import (
	"encoding/binary"
	"fmt"
	"net"
	"os"
	"sync"
	"syscall"
)

// SHMConn implements net.Conn using memory-mapped files for data transfer
// and standard TCP sockets for signaling event wakeups.
type SHMConn struct {
	net.Conn // the signaling connection
	mmapData []byte
	mmapFile *os.File
	mu       sync.Mutex
}

func NewSHMConn(signalConn net.Conn, shmPath string, size int, isOwner bool) (*SHMConn, error) {
	var file *os.File
	var err error

	if isOwner {
		file, err = os.OpenFile(shmPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
		if err != nil {
			return nil, err
		}
		if err = file.Truncate(int64(size)); err != nil {
			file.Close()
			return nil, err
		}
	} else {
		file, err = os.OpenFile(shmPath, os.O_RDWR, 0666)
		if err != nil {
			return nil, err
		}
	}

	data, err := syscall.Mmap(int(file.Fd()), 0, size, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
	if err != nil {
		file.Close()
		return nil, err
	}

	return &SHMConn{
		Conn:     signalConn,
		mmapData: data,
		mmapFile: file,
	}, nil
}

func (c *SHMConn) Read(b []byte) (int, error) {
	// Read 8-byte metadata header (offset, length) from the signaling socket
	header := make([]byte, 8)
	if _, err := ioReadFull(c.Conn, header); err != nil {
		return 0, err
	}

	offset := binary.BigEndian.Uint32(header[0:4])
	length := binary.BigEndian.Uint32(header[4:8])

	if int(offset+length) > len(c.mmapData) {
		return 0, fmt.Errorf("shm read out of bounds: offset=%d length=%d mmap_size=%d", offset, length, len(c.mmapData))
	}

	// Copy from shared memory diretamente into the destination slice
	n := copy(b, c.mmapData[offset:offset+length])
	return n, nil
}

func (c *SHMConn) Write(b []byte) (int, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	length := uint32(len(b))
	if int(length) > len(c.mmapData) {
		return 0, fmt.Errorf("data length exceeds shm capacity: %d > %d", length, len(c.mmapData))
	}

	// For simplicity, always write starting at offset 0 (ping-pong model per connection)
	const offset uint32 = 0

	// Write payload into memory mapping
	copy(c.mmapData[offset:offset+length], b)

	// Send notification (offset, length) via signaling socket
	header := make([]byte, 8)
	binary.BigEndian.PutUint32(header[0:4], offset)
	binary.BigEndian.PutUint32(header[4:8], length)

	if _, err := c.Conn.Write(header); err != nil {
		return 0, err
	}

	return len(b), nil
}

func (c *SHMConn) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	_ = syscall.Munmap(c.mmapData)
	if c.mmapFile != nil {
		_ = c.mmapFile.Close()
	}
	return c.Conn.Close()
}

// Helper to ensure full reads
func ioReadFull(r net.Conn, buf []byte) (int, error) {
	n := 0
	for n < len(buf) {
		read, err := r.Read(buf[n:])
		if err != nil {
			return n, err
		}
		n += read
	}
	return n, nil
}

// SHMListener wraps net.Listener to yield SHM connections
type SHMListener struct {
	net.Listener
	shmPath string
	shmSize int
}

func NewSHMListener(tcpListener net.Listener, shmPath string, shmSize int) *SHMListener {
	return &SHMListener{
		Listener: tcpListener,
		shmPath:  shmPath,
		shmSize:  shmSize,
	}
}

func (l *SHMListener) Accept() (net.Conn, error) {
	conn, err := l.Listener.Accept()
	if err != nil {
		return nil, err
	}
	return NewSHMConn(conn, l.shmPath, l.shmSize, true)
}

File: runtimes/rust/src/shm_transport.rs
#[cfg(unix)]
use std::fs::OpenOptions;
#[cfg(unix)]
use std::os::unix::io::AsRawFd;
use std::sync::Arc;
#[cfg(unix)]
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::Mutex;

pub struct ShmConn {
    pub signal: TcpStream,
    #[cfg(unix)]
    mmap_ptr: *mut libc::c_void,
    size: usize,
    write_mu: Arc<Mutex<()>>,
}

// Safely implement Send and Sync since access to mmap pointer is guarded/isolated
unsafe impl Send for ShmConn {}
unsafe impl Sync for ShmConn {}

impl ShmConn {
    #[cfg(unix)]
    pub fn new(
        signal: TcpStream,
        shm_path: &str,
        size: usize,
        is_owner: bool,
    ) -> Result<Self, std::io::Error> {
        let file = if is_owner {
            let f = OpenOptions::new()
                .read(true)
                .write(true)
                .create(true)
                .truncate(true)
                .open(shm_path)?;
            f.set_len(size as u64)?;
            f
        } else {
            OpenOptions::new().read(true).write(true).open(shm_path)?
        };

        let mmap_ptr = unsafe {
            libc::mmap(
                std::ptr::null_mut(),
                size,
                libc::PROT_READ | libc::PROT_WRITE,
                libc::MAP_SHARED,
                file.as_raw_fd(),
                0,
            )
        };

        if mmap_ptr == libc::MAP_FAILED {
            return Err(std::io::Error::last_os_error());
        }

        Ok(ShmConn {
            signal,
            mmap_ptr,
            size,
            write_mu: Arc::new(Mutex::new(())),
        })
    }

    #[cfg(not(unix))]
    pub fn new(
        _signal: TcpStream,
        _shm_path: &str,
        _size: usize,
        _is_owner: bool,
    ) -> Result<Self, std::io::Error> {
        Err(std::io::Error::new(
            std::io::ErrorKind::Unsupported,
            "Shared memory transport is not supported on Windows",
        ))
    }

    #[cfg(unix)]
    pub async fn read_payload(&mut self) -> Result<Vec<u8>, std::io::Error> {
        let mut header = [0u8; 8];
        self.signal.read_exact(&mut header).await?;

        let offset = u32::from_be_bytes([header[0], header[1], header[2], header[3]]) as usize;
        let length = u32::from_be_bytes([header[4], header[5], header[6], header[7]]) as usize;

        if offset + length > self.size {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "shm read out of bounds",
            ));
        }

        let mut data = vec![0u8; length];
        unsafe {
            let src = (self.mmap_ptr as *const u8).add(offset);
            std::ptr::copy_nonoverlapping(src, data.as_mut_ptr(), length);
        }

        Ok(data)
    }

    #[cfg(not(unix))]
    pub async fn read_payload(&mut self) -> Result<Vec<u8>, std::io::Error> {
        Err(std::io::Error::new(
            std::io::ErrorKind::Unsupported,
            "Shared memory transport is not supported on Windows",
        ))
    }

    #[cfg(unix)]
    pub async fn write_payload(&mut self, payload: &[u8]) -> Result<(), std::io::Error> {
        let _guard = self.write_mu.lock().await;

        let length = payload.len();
        if length > self.size {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "payload size exceeds shm capacity",
            ));
        }

        let offset = 0; // ping-pong model

        unsafe {
            let dest = (self.mmap_ptr as *mut u8).add(offset);
            std::ptr::copy_nonoverlapping(payload.as_ptr(), dest, length);
        }

        let mut header = [0u8; 8];
        header[0..4].copy_from_slice(&(offset as u32).to_be_bytes());
        header[4..8].copy_from_slice(&(length as u32).to_be_bytes());

        self.signal.write_all(&header).await?;
        self.signal.flush().await?;

        Ok(())
    }

    #[cfg(not(unix))]
    pub async fn write_payload(&mut self, _payload: &[u8]) -> Result<(), std::io::Error> {
        Err(std::io::Error::new(
            std::io::ErrorKind::Unsupported,
            "Shared memory transport is not supported on Windows",
        ))
    }
}

impl Drop for ShmConn {
    fn drop(&mut self) {
        #[cfg(unix)]
        unsafe {
            libc::munmap(self.mmap_ptr, self.size);
        }
    }
}
